pkg/tfbridge/provider_test.go

```diff
@@ -550,6 +521,104 @@ func TestProviderCheckV2(t *testing.T) {
 	assert.Equal(t, "", failures[2].Property)
 }
 
+func TestProviderCheck(t *testing.T) {
+	testFailures := map[string]func(*testing.T, []*pulumirpc.CheckFailure){
+		"v1": testCheckFailuresV1,
+		"v2": testCheckFailuresV2,
+	}
+	for _, f := range factories {
+		provider := &Provider{
+			tf:     f.NewTestProvider(),
+			config: f.NewTestProvider().Schema(),
+		}
+
+		provider.resources = map[tokens.Type]Resource{
+			"SecondResource": {
+				TF:     provider.tf.ResourcesMap().Get("second_resource"),
+				TFName: "second_resource",
+				Schema: &ResourceInfo{Tok: "SecondResource"},
+			},
+		}
+
+		t.Run(f.SDKVersion(), func(t *testing.T) {
+			t.Run("failures", func(t *testing.T) {
+				failures := testCheckFailures(t, provider, "SecondResource")
+				assert.Len(t, failures, 3)
+				sort.SliceStable(failures, func(i, j int) bool {
+					return failures[i].Reason < failures[j].Reason
+				})
+				testFailures[f.SDKVersion()](t, failures)
+			})
+		})
+	}
+}
+
+func TestCheckCallback(t *testing.T) {
+	test := func(t *testing.T, p *Provider) {
+		testutils.ReplaySequence(t, p, `
+			[
+			  {
+			    "method": "/pulumirpc.ResourceProvider/Configure",
+			    "request": {
+			      "args": {
+				"prop": "global"
+			      },
+			      "variables": {
+				"prop": "global"
+			      }
+			    },
+			    "response": {
+			      "supportsPreview": true
+			    }
+			  },
+			  {
+			    "method": "/pulumirpc.ResourceProvider/Check",
+			    "request": {
+			      "urn": "urn:pulumi:st::pg::testprovider:index/res:Res::r",
+			      "olds": {},
+			      "news": {},
+			      "randomSeed": "wqZZaHWVfsS1ozo3bdauTfZmjslvWcZpUjn7BzpS79c="
+			    },
+			    "response": {
+			      "inputs": {
+				"__defaults": [],
+				"arrayPropertyValues": ["global"]
+			      }
+			    }
+			  }
+			]`)
+	}
+
+	callback := func(
+		_ context.Context, config, meta resource.PropertyMap,
+	) (resource.PropertyMap, error) {
+		config["arrayPropertyValues"] = resource.NewArrayProperty(
+			[]resource.PropertyValue{meta["prop"]},
+		)
+		return config, nil
+	}
+
+	for _, f := range factories {
```

Comment:
By-reference loop variable capture :)

Reply:
The test isn't parallelized. I can change that (https://github.com/pulumi/pulumi-terraform-bridge/commit/272e881f24744a86655da45c59f5c903d5c0de46), but the by-ref isn't actually a risk here.

Reply:
Yeah I wanted to enable a linter that flags these but we can deal with it then.

Reply:
Have you seen https://github.com/golang/go/discussions/56010?